protocol/messenger.go

```diff
@@ -1778,88 +1785,161 @@ func (m *Messenger) handlePushNotificationClientRegistrations(c chan struct{}) {
 // InitFilters analyzes chats and contacts in order to setup filters
 // which are responsible for retrieving messages.
 func (m *Messenger) InitFilters() error {
-
 	// Seed the for color generation
 	rand.Seed(time.Now().Unix())
 
-	logger := m.logger.With(zap.String("site", "Init"))
-
 	// Community requests will arrive in this pubsub topic
-	err := m.SubscribeToPubsubTopic(shard.DefaultNonProtectedPubsubTopic(), nil)
+	if err := m.SubscribeToPubsubTopic(shard.DefaultNonProtectedPubsubTopic(), nil); err != nil {
+		return err
+	}
+
+	filters, publicKeys, err := m.collectFiltersAndKeys()
 	if err != nil {
 		return err
 	}
 
-	var (
-		filtersToInit []transport.FiltersToInitialize
-		publicKeys    []*ecdsa.PublicKey
-	)
+	_, err = m.transport.InitFilters(filters, publicKeys)
+	return err
+}
+
+func (m *Messenger) collectFiltersAndKeys() ([]transport.FiltersToInitialize, []*ecdsa.PublicKey, error) {
+	var wg sync.WaitGroup
+	errCh := make(chan error, 5)
+	filtersCh := make(chan []transport.FiltersToInitialize, 3)
+	publicKeysCh := make(chan []*ecdsa.PublicKey, 2)
+
+	wg.Add(5)
+	go m.processJoinedCommunities(&wg, filtersCh, errCh)
+	go m.processSpectatedCommunities(&wg, filtersCh, errCh)
+	go m.processChats(&wg, filtersCh, publicKeysCh, errCh)
+	go m.processContacts(&wg, publicKeysCh, errCh)
+	go m.processControlledCommunities(&wg, errCh)
+
+	wg.Wait()
+	close(filtersCh)
+	close(publicKeysCh)
+
+	select {
+	case err := <-errCh:
+		return nil, nil, err
+	default:
+	}
+
+	return m.collectResults(filtersCh, publicKeysCh)
```

Comment:
I think `errCh` can also be passed to `collectResults` and processed inside.
Right now `collectResults` always returns `nil` error.